PyInputToolExample/engine.py

```diff
@@ -0,0 +1,210 @@
+            raise
+
+        #
+        # the Engine is ready for us to get the Output anchor. We know it is called 'Output' because
+        #  that is what we put in our Config.xml as its name
+        #
+
+        self.output_anchor = self.output_anchor_mgr.get_output_anchor('Output')
+
+        self.initialized = True
+
+        return
+
+    #
+    # pi_close will be called after all the records have been processed
+    #
+    def pi_close(self, b_has_errors):
+        self.closed = True
+        self.output_anchor.close()
+        return
+
+    #
+    # pi_add_incoming_connection will be called when a new input is connected to this tool
+    #
+    def pi_add_incoming_connection(self, str_type, str_name):
+        self.output_message(
+            'pi_add_incoming_connection'
+            , AlteryxPythonSDK.EngineMessageType.error
+            , 'This tool does not accept an Incoming Connection'
+        )
+        return self
+
+    #
+    # pi_add_outgoing_connection will be called when a new output is connected to this tool
+    #
+    def pi_add_outgoing_connection(self, str_name):
+        return True
+
+    #
+    # pi_push_all_records will be called if there are no inputs connected to this tool
+    #
+    def pi_push_all_records(self, n_record_limit):
+        if self.initialized != True:
+            return False
+
+        # stop processing and returns error if the file is not a csv
+        if self.is_csv() != True:
+            self.output_message('Error', AlteryxPythonSDK.EngineMessageType.error, 'This tool only accepts csv files')
+            return False
+
+        self.record_info_out = AlteryxPythonSDK.RecordInfo()
+        # create a read-only file object
+        self.file_out = open(self.file_input_name, 'r')
+        # map the information read into a dict where the fieldnames are the keys
+        self.file_reader = csv.DictReader(self.file_out)
+
+        #
+        # add metadata info that is passed to tools downstream
+        #
+        for field in self.file_reader.fieldnames:
+            self.record_info_out.add_field(
+                field
+                , AlteryxPythonSDK.FieldType.v_string
+                , 254
+                , 0
+                , self.name
+                , ''
+            )
+
+        #
+        # tell the downstream tools what our output will look like
+        #
+        self.output_anchor.init(self.record_info_out, '')
+
+        #
+        # create the helper for constructing records to pass downstream
+        #
+        self.record_creator = self.record_info_out.construct_record_creator()
+        rownum = 0
+
+        #
+        # map values in file to the tool output
+        #
+        for row in self.file_reader:
+            rownum += 1
+            for index, value in enumerate(row.items()):
+                self.record_info_out[index].set_from_string(self.record_creator, value[1])
```

Comment:
value consists of (fieldname, value) at the specified index, so value[0] is the fieldname